### PR TITLE
Correct carry's interval

### PR DESCRIPTION
**Pull request by Nevrdid**
Disclaim :
When apply for first time, alive sourcers will use a default path's terrain datas of 10 plain. The correct values will be calculated by next sourcers spawned.
It may slow down a little/alot while waiting for next sourcers as it may be better to suicide sourcers in base if you have energy for respawn them.

### CODE REVIEW

File: src/prototype_creep_harvest.js [hunk 1]
```diff
@@ -59,9 +59,9 @@ Creep.prototype.spawnCarry = function() {
   let carryCapacity = config.carry.sizes[foundKey][1] * CARRY_CAPACITY;
 
   const workParts = this.body.filter(part => part.type === WORK).length;
-
-  let waitTime = carryCapacity / (HARVEST_POWER * workParts);
-
+  let distance = Game.map.getRoomLinearDistance(this.room.name, this.memory.base);
+  let waitTime = Math.floor(1500 * carryCapacity * distance / (2 * this.memory.timeToTravel * (HARVEST_POWER * workParts))) ;
```
Comment (Nevrdid) [inline src/prototype_creep_harvest.js]: Time between carrys should be proportional to % of energy the carry will carry in his life as : 
Energy harvested in sourcer life :  
`1500 * (HARVEST_POWER * workParts) = A`
Energy carried by carry : 
`1500 * carryCapacity / 2 * timetotravel = B`
`B/A = carryCapacity /( 2 * timetotravel * harvestpower * workParts) `
This result should be interpreted as spawn a creep each 1500 ticks is B/A = 1. One each 750 ticks if it's 1/2.
then we just have to multiply it by 1500 and floor all for have a rounded value.  
The distance coeff is for take care of surpopulation in parents rooms, then far rooms will call less carry.

File: src/prototype_creep_routing.js [hunk 2]
```diff
@@ -325,6 +326,11 @@ Creep.prototype.moveByPathMy = function(route, routePos, start, target, skipPreM
 
   this.move(directions.direction);
 
+  if (this.isStuck()) {
+    this.say('...');
+    this.moveCreep(path[pathPos + 1], directions.direction > 4  ? directions.direction - 4 : directions.direction + 4);
+  }
+
```
Comment (Nevrdid) [inline src/prototype_creep_routing.js]: gonna remove this ^^'

File: src/prototype_creep_harvest.js (hunk 1 above)
Comment (kahojyun) [reply src/prototype_creep_harvest.js]: take `resourceAtPosition` into account would be better.

**Comment by Nevrdid** [comment]
Updated. working well on my private. If anyone with multiple rooms can try it for say me if anything happen badly it could be great :)

File: src/prototype_creep_harvest.js [hunk 3]
```diff
@@ -62,16 +64,50 @@ Creep.prototype.spawnCarry = function() {
...
     }
   }
-  this.memory.wait = Math.max(waitTime, config.carry.minSpawnRate);
+  this.memory.wait = this.getCarrySpawnInterval(parts, resourceAtPosition);
+};
+
+/*
+Time between carrys should be proportional to % of energy the carry will carry in his life as :
+Energy harvested in sourcer life :
+`1500 * (HARVEST_POWER * workParts) = A`
```
Comment (samogot) [inline src/prototype_creep_harvest.js]: replace 1500 with `CREEP_LIFE_TIME`. also in code

File: src/prototype_creep_harvest.js [hunk 4]
```diff
@@ -62,16 +64,50 @@ Creep.prototype.spawnCarry = function() {
...
+Energy carried by carry :
+`1500 * carryCapacity / carryTravelTime = B`
+`B/A = carryCapacity /( carryTravelTime * harvestpower * workParts) `
+This result should be interpreted as spawn a creep each 1500 ticks is B/A = 1. One each 750 ticks if it's 1/2.
+then we just have to multiply it by 1500 and floor all for have a rounded value.
+The distance coeff is for take care of surpopulation in parents rooms, then far rooms will call less carry. Not even sure we need it.
```
Comment (samogot) [inline src/prototype_creep_harvest.js]: surpopulation?

File: src/prototype_creep_harvest.js [hunk 5]
```diff
@@ -62,16 +64,50 @@ Creep.prototype.spawnCarry = function() {
...
     });
     if (nearCarries.length > 1) {
       nearCarries[0].memory.recycle = true;
     }
   }
-  this.memory.wait = Math.max(waitTime, config.carry.minSpawnRate);
+  this.memory.wait = this.getCarrySpawnInterval(parts, resourceAtPosition);
+};
+
+/*
```
Comment (samogot) [inline src/prototype_creep_harvest.js]: make it valid JSDoc please

File: src/prototype_creep_harvest.js [hunk 6]
```diff
@@ -62,16 +64,50 @@ Creep.prototype.spawnCarry = function() {
...
+  let terrain;
+  const terrainCost = {plain: 1, swamp: 5, road: 0.5};
+  for (terrain of ['plain', 'swamp', 'road']) {
+    travelTime += this.memory.pathDatas[terrain] * Math.max(1, terrainCost[terrain] * Math.ceil(carryParts.work / carryParts.move));
+    travelTime += this.memory.pathDatas[terrain] * Math.max(1, terrainCost[terrain] * Math.ceil((carryParts.work + carryParts.carry) / carryParts.move));
+  }
+
+  const sourceFullFilled = Game.getObjectById(this.memory.routing.targetId).energyCapacity / 3000;
+  // const distance = this.getRoute(this.room.name, this.memory.base).length || Game.map.getRoomLinearDistance(this.room.name, this.memory.base);
+  const waitTime = Math.floor(1500 * carryParts.carry * CARRY_CAPACITY /* * distance */ / ((sourceFullFilled * HARVEST_POWER * sourcerWork + resourcesDroped) * travelTime));
```
Comment (samogot) [inline src/prototype_creep_harvest.js]: add info about `sourceFullFilled` and `resourcesDroped` into formula in docs comment. I guess it should be `resourcesDroped/CREEP_LIFE_TIME` as we want to get rid of dropped resources in one creep live, not in one tick

File: src/prototype_creep_harvest.js [hunk 7]
```diff
@@ -62,16 +64,50 @@ Creep.prototype.spawnCarry = function() {
...
+  const spawnTime = _.sum(carryParts) * CREEP_SPAWN_TIME;
+  // return Math.max(waitTime, config.carry.minSpawnRate);
+  return Math.max(waitTime, spawnTime);
```
Comment (samogot) [inline src/prototype_creep_harvest.js]: I think `travelTime` should be here as one of max parameters. We should not call any creep, if first one called haven't achieved us yet, and haven't decreased amount of dropped resources

File: src/prototype_creep_harvest.js [hunk 8]
```diff
@@ -62,16 +64,50 @@ Creep.prototype.spawnCarry = function() {
...
+  let travelTime = 0;
+
+  let terrain;
+  const terrainCost = {plain: 1, swamp: 5, road: 0.5};
+  for (terrain of ['plain', 'swamp', 'road']) {
+    travelTime += this.memory.pathDatas[terrain] * Math.max(1, terrainCost[terrain] * Math.ceil(carryParts.work / carryParts.move));
+    travelTime += this.memory.pathDatas[terrain] * Math.max(1, terrainCost[terrain] * Math.ceil((carryParts.work + carryParts.carry) / carryParts.move));
+  }
+
+  const sourceFullFilled = Game.getObjectById(this.memory.routing.targetId).energyCapacity / 3000;
```
Comment (TooAngel) [inline src/prototype_creep_harvest.js]: If you want to take a source metric into account you need to compare it to the time to regenerate. `energyCapacity = 10` is totally fine if it takes only 1 tick to be filled up again.

**Comment by TooAngel** [comment]
Hm, shouldn't have been merged. Fixed in democratic-collaboration https://github.com/TooAngel/democratic-collaboration/pull/77

**Pull request merged**